### DISCUSSION TITLE
Airflow Listeners vs Callbacks

### DISCUSSION THREAD

**Discussion by asluis3** [apache/airflow#33238] [General] [answered]
Hi everyone,

I am looking through the Airflow documentation and I cannot find the difference between callbacks and listeners. Could someone explain why both co-exist and what benefits one might have over the other? The main difference I found is that listeners operate as  a plugin whereas callbacks are native to Airflow. Thank you!

**For reference**
**Callback docs:** https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/callbacks.html
**Listener docs:** https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/listeners.html

**Comment by potiuk** [comment, marked as answer]
* Callbacks/tasks are declared in each task separately and are per-task defined by DAG authors.
* Listenaers are declared once for evyrything and not by DAG authors but Deployment Manager

It has hudge implications also when it comes to Airflow's security model:  https://airflow.apache.org/docs/apache-airflow/stable/security/index.html#airflow-security-model-user-types

**Reply by asluis3** [reply]
Hi @potiuk, do you happen to know which component in Airflow's architecture does the Listener run in? Would it be in the  scheduler like most other plugins? I'm not able to find more information on that—thank you.

**Reply by potiuk** [reply]
I think it depends on listener and is bound to event type - job listeners are where job is run (so depends what type of job it is) task ones are done when the tasks are running. But this might be good to explain it in the docs @mobuchowski .
